source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -1342,6 +1236,125 @@ bool COpenGLDriver::beginScene(bool backBuffer, bool zBuffer, SColor color,
+    spvCPUShader->drop();
+    return core::smart_refctd_ptr<COpenGLSpecializedShader>(gpuSpecialized, core::dont_grab);
+}
+
+core::smart_refctd_ptr<IGPUBufferView> COpenGLDriver::createGPUBufferView(IGPUBuffer* _underlying, asset::E_FORMAT _fmt, size_t _offset, size_t _size)
+{
+    if (!_underlying)
+        return nullptr;
+    const size_t effectiveSize = (_size != IGPUBufferView::whole_buffer) ? (_underlying->getSize() - _offset) : _size;
+    if ((_offset + effectiveSize) > _underlying->getSize())
+        return nullptr;
+    if ((_offset / reqTBOAlignment)*reqTBOAlignment != _offset) //offset must be aligned to GL_TEXTURE_BUFFER_OFFSET_ALIGNMENT
+        return nullptr;
+    if (!isAllowedBufferViewFormat(_fmt))
+        return nullptr;
+    if (effectiveSize > (maxTBOSizeInTexels * asset::getTexelOrBlockBytesize(_fmt)))
+        return nullptr;
+
+    COpenGLBuffer* glbuf = static_cast<COpenGLBuffer*>(_underlying);
+    return core::make_smart_refctd_ptr<COpenGLBufferView>(core::smart_refctd_ptr<COpenGLBuffer>(glbuf), _fmt, _offset, _size);
+}
+
+core::smart_refctd_ptr<IGPUDescriptorSetLayout> COpenGLDriver::createGPUDescriptorSetLayout(const IGPUDescriptorSetLayout::SBinding* _begin, const IGPUDescriptorSetLayout::SBinding* _end)
+{
+    return core::make_smart_refctd_ptr<IGPUDescriptorSetLayout>(_begin, _end);//there's no COpenGLDescriptorSetLayout (no need for such)
+}
+
+core::smart_refctd_ptr<IGPUSampler> COpenGLDriver::createGPUSampler(const IGPUSampler::SParams& _params)
+{
+    return core::make_smart_refctd_ptr<COpenGLSampler>(_params);
+}
+
+core::smart_refctd_ptr<IGPUPipelineLayout> COpenGLDriver::createGPUPipelineLayout(const asset::SPushConstantRange* const _pcRangesBegin, const asset::SPushConstantRange* const _pcRangesEnd, core::smart_refctd_ptr<IGPUDescriptorSetLayout>&& _layout0, core::smart_refctd_ptr<IGPUDescriptorSetLayout>&& _layout1, core::smart_refctd_ptr<IGPUDescriptorSetLayout>&& _layout2, core::smart_refctd_ptr<IGPUDescriptorSetLayout>&& _layout3)
+{
+    return core::make_smart_refctd_ptr<COpenGLPipelineLayout>(
+        _pcRangesBegin, _pcRangesEnd,
+        std::move(_layout0), std::move(_layout1),
+        std::move(_layout2), std::move(_layout3)
+        );
+}
+
+core::smart_refctd_ptr<IGPURenderpassIndependentPipeline> COpenGLDriver::createGPURenderpassIndependentPipeline(core::smart_refctd_ptr<IGPUPipelineLayout>&& _layout, core::smart_refctd_ptr<IGPUSpecializedShader>&& _vs, core::smart_refctd_ptr<IGPUSpecializedShader>&& _tcs, core::smart_refctd_ptr<IGPUSpecializedShader>&& _tes, core::smart_refctd_ptr<IGPUSpecializedShader>&& _gs, core::smart_refctd_ptr<IGPUSpecializedShader>&& _fs, const asset::SVertexInputParams& _vertexInputParams, const asset::SBlendParams& _blendParams, const asset::SPrimitiveAssemblyParams& _primAsmParams, const asset::SRasterizationParams& _rasterParams)
```

Comment:
maybe just instead of 1 pointer per each shader type, accept an array (or begin/end) of pair<core::smart_refctd_ptr<IGPUSpecializedShader>,E_SHADER_STAGE> by r-value or something?

There will be more shader types in Vulkan (mesh and raytracing)

Reply:
same thing as https://github.com/buildaworldnet/IrrlichtBAW/pull/358#discussion_r332878486